core/src/main/java/org/apache/iceberg/SerializedTable.java

```diff
@@ -0,0 +1,312 @@
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg;
+
+import java.io.Serializable;
+import java.util.List;
+import java.util.Map;
+import org.apache.iceberg.encryption.EncryptionManager;
+import org.apache.iceberg.hadoop.HadoopFileIO;
+import org.apache.iceberg.hadoop.SerializedConfiguration;
+import org.apache.iceberg.io.FileIO;
+import org.apache.iceberg.io.LocationProvider;
+import org.apache.iceberg.util.SerializableMap;
+
+/**
+ * A read-only serializable table implementation that can be sent to other nodes.
+ * <p>
+ * While this class captures the metadata file location that can be used to load the full table
+ * metadata, it also serializes directly things like properties, schema, spec, sort order
+ * to avoid reading the metadata file from other nodes to access frequently needed metadata.
+ * <p>
+ * <em>Note:</em> loading the full metadata from a large number of nodes can overwhelm the storage.
+ */
+public class SerializedTable implements Table, Serializable {
+  private final String name;
+  private final String location;
+  private final String metadataFileLocation;
+  private final Map<String, String> properties;
+  private final String schemaAsJson;
+  private final String specAsJson;
+  private final String sortOrderAsJson;
+  private final FileIO io;
+  private final EncryptionManager encryption;
+  private final LocationProvider locationProvider;
+
+  private transient volatile Table lazyTable = null;
+  private transient volatile Schema lazySchema = null;
+  private transient volatile PartitionSpec lazySpec = null;
+  private transient volatile SortOrder lazySortOrder = null;
+
+  public SerializedTable(Table table) {
+    this.name = table.name();
+    this.location = table.location();
+    this.metadataFileLocation = metadataFileLocation(table);
+    this.properties = SerializableMap.copyOf(table.properties());
+    this.schemaAsJson = SchemaParser.toJson(table.schema());
+    this.specAsJson = PartitionSpecParser.toJson(table.spec());
+    this.sortOrderAsJson = SortOrderParser.toJson(table.sortOrder());
+    this.io = fileIO(table);
+    this.encryption = table.encryption();
+    this.locationProvider = table.locationProvider();
+  }
+
+  private String metadataFileLocation(Table table) {
+    if (table instanceof HasTableOperations) {
+      TableOperations ops = ((HasTableOperations) table).operations();
+      return ops.current().metadataFileLocation();
+    } else {
+      return null;
+    }
+  }
+
+  private FileIO fileIO(Table table) {
```

Comment:
I've changed this place to handle `FileIO` instead of delegating it to the caller.
We rely on `SerializedConfiguration` that is introduced in this PR.

Reply:
This can be made more generic. We can expose `ConfigurableFileIO` interface or something similar instead of only handling `HadoopFileIO`.

Reply:
We can do something similar to dynamic loading, use `if (table.io() instanceof Configurable)`

Reply:
Yeah, we would need a way to construct a copy of the class with the serialized conf.

If we knew that all `FileIO` implementations can be dynamically loaded, we could simply persist the class, props, and optional conf and rebuild it on demand. However, we cannot guarantee all `FileIO` implementations are dynamically loaded.

Reply:
Ah I see, the FIleIO has to directly accept the serialized hadoop config supplier. I thought it's enough after #2333, but the `setConf()` there still needs to accept a Hadoop config and would not work with the supplier.

In that case, I think we can potentially have something like a `SerializedConfigurable` interface that has method `setSerializedConfSupplier(supplier)`, and check instance of that class. `HadoopFileIO` can then implement that class instead.

The main point here is that if we do it for HadoopFileIO only, I think we should at least make sure other FileIOs that leverage hadoop configuration can work with this.

Reply:
Yeah, we would definitely need a more generic solution for `FileIO` implementations that depend on Hadoop conf. 
I tried to prototype that but it raised a number of questions. The last idea I had was something like this:

```
interface ConfigurableFileIO<T extends ConfigurableFileIO> extends FileIO, Configurable {
  T toSerializable(SerializedConfiguration serializedConf);

  default Object writeReplace() {
    // default the implementation to call `toSerializable`
  }
}
```

It was clear this requires a separate PR and more thinking. That's why I propose to address that in a follow-up.

Reply:
Cool, let's discuss in another PR then. Just to throw another idea I just have:

```java
public interface KryoSerializable<T> extends Serializable {
  
  // return a serialized version of self
  default T serialized() {
    throw new UnsupportedOperationException("Cannot support kryo serialization");
  }
}
```

```java
public interface FileIO extends KryoSerializable<FileIO> {
 ...
}
```

```java
public class HadoopFileIO ... {
  @Override
  public FileIO serialized() {
    SerializedConfiguration serializedConf = new SerializedConfiguration(getConf());
    return new HadoopFileIO(serializedConf::get);
  }
}
```

By doing this, we enforce everything dynamically loaded to be kryo serializable.

Reply:
I don't think that we should change the contract for dynamically loaded classes. If a user chooses to use Kryo and a dynamically loaded component, it is their responsibility to make sure the two are compatible. We just need to make sure that Iceberg-supplied classes work with Kryo.